### DISCUSSION TITLE
Control over Avalonia Message Loop

### DISCUSSION THREAD

**Discussion by kuiperzone** [AvaloniaUI/Avalonia#17023] [Q&A]
It would be complex to explain fully what I'm trying to do here, but I need to chain together custom blocking functions called in response to events, while preserving the MainWindow  and responding to Avalonia events.

Essentially,  I need to enter the Avalonia message loop, come out of it on demand, and then go back in.

_To phrase it another way, I need the equivalent of a `Window.ShowDialog()` method on a Window that is already visible._

I'm reading:
https://docs.avaloniaui.net/docs/concepts/application-lifetimes

I see the code:

```
  // Application entry point. Avalonia is completely initialized.
  static void AppMain(Application app, string[] args)
  {
     // A cancellation token source that will be 
     // used to stop the main loop
     var cts = new CancellationTokenSource();
     
     // Do your startup code here
     new Window().Show();

     // Start the main loop
     app.Run(cts.Token);
  }
```

Q. What I'm asking is can I cancel `Run()`, end then re-enter it? Something like...

     new Window().Show();

     var cts1 = new CancellationTokenSource();
     app.Run(cts1.Token);

     // Cancelled, but re-enter
     var cts2 = new CancellationTokenSource();
     app.Run(cts2.Token);

Would this work?

Is the another way to do it?

**Comment by maxkatz6** [comment]
You need a `DispatcherFrame` API, which is more or less the same in Avalonia like on WPF.
Keep in mind it will only work on Desktop where we have full control over app loop. We don't have that on mobile or browser (normally).

**Comment by kuiperzone** [comment]
Just want to thank you for your reply. I may follow this up when I have discovered a path to go down for my solution.

**Comment by kuiperzone** [comment]
I'm not up to speed on WASM yet, but would need to my solution to eventually work in a browser.

I'm thinking that the way to go might be to employ 2 long running threads. One is the app main which can handle Avalonia. The other I will use to handle the blocking calls as mentioned above. The only thing needed would be a way to get the two to communicate.

Q. I'm a little confused by threads in WASM. In some literature, I read that they are not implemented in .NET 8 but might be .NET 9.

On the other hand, I see you guys already using them:

https://github.com/AvaloniaUI/Avalonia/issues/14022

Any problems in WASM using the Thread class or Task?

**Reply by maxkatz6** [reply]
.NET 8 WASM supports threads (including Thread class), but .NET 9 has it completely refactored and redone on runtime level.
In case of Avalonia, we supported .NET 8 WASM threads, but we are going to only support .NET 9.

Specifically about combination of Avalonia 11.2 and .NET 9 WASM threading, `DispatcherFrame` API will work there.

**Reply by kuiperzone** [reply]
Thank you @maxkatz6
